test/common/utils.cpp

```diff
@@ -148,3 +148,158 @@ TEST_CASE("sha256_of() works as expected") {
+        auto maybe_res =
+            mk::slurpv_impl<uint8_t>("./test/fixtures/text-with-utf8.txt");
+        REQUIRE(!!maybe_res);
+        std::vector<uint8_t> expect{
+            'C', 'i', 'r', 'i',  0xc3, 0xa8, '\n',      // first line
+            'A', 'g', 'l', 'i',  0xc3, 0xa8, '\n',      // second line
+            'P', 'e', 'l', 0xc3, 0xa8, '\n',            // third line
+            'P', 'e', 'r', 'c',  'h',  0xc3, 0xa9, '\n' // fourh line
+        };
+        REQUIRE(*maybe_res == expect);
+        REQUIRE(maybe_res->size() == 28);
+    }
+
+    SECTION("For a purely binary file") {
+        auto maybe_res = mk::slurpv_impl<uint8_t>("./test/fixtures/gzipped.gz");
+        REQUIRE(!!maybe_res);
+        std::vector<uint8_t> expect{
+            0x1f, 0x8b, 0x08, 0x08, 0xad, 0x82, 0x4d, 0x58,
+            0x02, 0x03, 0x58, 0x4f, 0x00, 0xad, 0xce, 0xb1,
+
+            0x4e, 0xc3, 0x30, 0x10, 0xc6, 0xf1, 0xdd, 0x4f,
+            0x71, 0x23, 0x20, 0x5d, 0x1c, 0x40, 0x62, 0x88,
+
+            0xc4, 0x80, 0x1a, 0x50, 0x17, 0x28, 0x82, 0x6c,
+            0x88, 0xc1, 0xb1, 0x8f, 0xc6, 0xc8, 0xb1, 0x8d,
+
+            0xef, 0x0a, 0xe2, 0xed, 0x69, 0x52, 0x44, 0x79,
+            0x80, 0xdc, 0xfa, 0x49, 0xff, 0xdf, 0x75, 0x9b,
+
+            0x76, 0xa3, 0xae, 0xf7, 0xa7, 0x54, 0x37, 0x10,
+            0x13, 0x08, 0xb1, 0x30, 0xb0, 0xf8, 0x10, 0x20,
+
+            0x12, 0x39, 0x90, 0x04, 0x3d, 0x81, 0x1f, 0x73,
+            0xa0, 0x91, 0xa2, 0x90, 0x6b, 0x94, 0x3a, 0x83,
+
+            0x97, 0xf6, 0xe1, 0x19, 0x56, 0x29, 0xb2, 0x67,
+            0xa1, 0x68, 0xbf, 0x5f, 0x4f, 0x06, 0x91, 0xcc,
+
+            0x8d, 0xd6, 0x5b, 0x2f, 0x5f, 0xd4, 0x57, 0x92,
+            0x4a, 0x2e, 0xe9, 0x9d, 0xac, 0x54, 0xa9, 0x6c,
+
+            0x75, 0x4a, 0xd1, 0x6b, 0xce, 0x64, 0xab, 0xfd,
+            0xae, 0xfb, 0x90, 0x7a, 0xbd, 0xbe, 0xbd, 0x69,
+
+            0x1b, 0x3d, 0x79, 0x38, 0x0d, 0xac, 0x85, 0xb1,
+            0xae, 0x2f, 0xd0, 0x45, 0xb6, 0xc7, 0x70, 0x35,
+
+            0xba, 0xd3, 0x19, 0x5c, 0x77, 0xdd, 0x23, 0x3c,
+            0xd1, 0xc7, 0x6e, 0x7a, 0x70, 0x31, 0xee, 0x12,
+
+            0xa7, 0x10, 0x96, 0xdf, 0xee, 0x51, 0x23, 0xe3,
+            0xa8, 0xc0, 0x9d, 0xa7, 0xe0, 0xe0, 0xde, 0x44,
+
+            0x9f, 0x77, 0xc1, 0x88, 0x4f, 0x71, 0x31, 0xf9,
+            0x0a, 0x87, 0xd9, 0xc0, 0xb7, 0xc9, 0xc0, 0xf1,
+
+            0x9f, 0xf1, 0xf7, 0xc5, 0xca, 0x64, 0xf1, 0x9f,
+            0x04, 0x39, 0x15, 0x31, 0x61, 0x29, 0xfa, 0xbc,
+
+            0x46, 0x7b, 0x08, 0xe3, 0x21, 0x3c, 0x7b, 0x3f,
+            0x31, 0x51, 0x2e, 0xef, 0x0a, 0x02, 0x00, 0x00,
+        };
+        REQUIRE(*maybe_res == expect);
+        REQUIRE(maybe_res->size() == 240);
+    }
+}
+
+TEST_CASE("slurp() works as expected") {
+    SECTION("In case of nonexistent file") {
+        auto maybe_res = mk::slurp("/nonexistent");
+        REQUIRE(!maybe_res);
+        REQUIRE(maybe_res.as_error().code == mk::FileIoError().code);
```

Comment:
Back in the day `Error` automatically converted to `int`. Then came `ErrorOr<T>` with automatic constructors for both `Error` and `T`. Then I started to wonder what would happen if `ErrorOr<int>` was passed a `Error`. Could the compiler, by mistake, convert the error to `int`?

This scared me and I reacted emotionally by removing automatic conversion to `int`. Probably it was not the optimal, rational reaction and it would have been better to add unit tests for that. I think we are still in time to amend my previous behavior and restore `Error` to its ancient glory.

Until we do that, we have the minor issue that `Catch` passes things you put into `REQUIRE` clauses to `std::cout` (or a similar stream). Unfortunately `Error` is only convertible to `bool`, therefore when `REQUIRE(err == GenericError())` is `false` you see something like:

> Assertion failed `1 == 1`

Which is counter-intuitive to say the least. That why I started adding `.code` explicitly. The same problem applies for the places in which I print `Error` using `%d`.

Now, considering that one can always convert to integer through a boolean (as is clearly exemplified in the `%d` case), I am not even sure my original fix was wise because conversion to `int` would probably still be possible through boolean. Furthermore, thinking more about the matter, I do not think a compiler would be allowed to implicitly convert `Error` to `int` when a constructor exists for both, given that `Error` is the type and `int` is available only after applying the cast operator.

Anyway, I guess we should open a issue about this and move on. Interesting how I did not see the matter until you commented and then, one second after reading the comment, the whole thing &mdash; including the wishful thinking, poor man fix &mdash; was apparent.